File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> result;
    void twoSum(vector<int>&nums,int target ,int a, int b){
        int n = nums.size();
        
        while(a<b){
            if(nums[a] + nums[b] < target ){
                a++;
            }else if(nums[a] + nums[b] > target){
                b--;
            }else{
                while(a<b && nums[a] == nums[a+1]) a++;
                while(a<b && nums[b] == nums[b-1]) b--;
                result.push_back({ -target , nums[a],nums[b]});

                a++;
                b--;
            }
           


        }
    }

    
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();

        if(n<3) return {};

        result.clear();

        sort(begin(nums),end(nums));

        for(int i=0 ; i<n ; i++){
            if(i>0 && nums[i] == nums[i-1]){
                continue;
            }
            int n1 = nums[i];
            int target = -n1;

            twoSum(nums, target , i+1 , n-1);
        }    
        return result;    
    }
};

File: 0015-3sum/0015-3sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);

        List<List<Integer>> ans = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {

            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int a = i + 1;
            int b = nums.length - 1;

            while (a < b) {
                int sum = nums[i] + nums[a] + nums[b];

                if (sum == 0) {
                    ans.add(Arrays.asList(nums[i], nums[a], nums[b]));

                    a++;
                    b--;

                    while (a < b && nums[a] == nums[a - 1]) {
                        a++;
                    }

                    while (a < b && nums[b] == nums[b + 1]) {
                        b--;
                    }

                } else if (sum < 0) {
                    a++;
                } else {
                    b--;
                }
            }
        }

        return ans;
    }
}